File: 백준/12865.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n').map((data) => data.split(' ').map(Number));
let [N, K] = input[0];
let w = new Array(N + 1).fill(0);
let v = new Array(N + 1).fill(0);
let dp = new Array(K + 1).fill(0);

for (let i = 1; i <= N; i++) {
    let [wInput, vInput] = input[i];
    w[i] = wInput;
    v[i] = vInput;
}

for (let i = 1; i <= N; i++) { // 물품 index
    for (let j = K; j >= 1; j--) { // dp index
        if (w[i] <= j)
            dp[j] = Math.max(dp[j], dp[j - w[i]] + v[i]);
    }
}

console.log(dp[K]);

File: 백준/12865.py
import sys

sys.stdin = open("data.txt", 'r')

input = sys.stdin.readline

N, K = map(int, input().split(" "))
items = [(0, 0)]

for _ in range(N):
    w, v = map(int, input().split(" "))
    items.append((w, v))


def solution(N, K, items):
    dp = [[0] * (K + 1) for _ in range(N + 1)]

    for i in range(1, N + 1):
        weight, value = items[i]

        for j in range(1, K + 1):
            if j < weight:
                dp[i][j] = dp[i - 1][j]
            else:
                dp[i][j] = max(dp[i - 1][j], value + dp[i - 1][j - weight])

    print(dp[N][K])
    return


solution(N, K, items)
